app/scripts/metamask-controller.js

```diff
@@ -1562,39 +1555,98 @@ module.exports = class MetamaskController extends EventEmitter {
   }
 
   /**
-   * A method for providing our public api over a stream.
-   * This includes a method for setting site metadata like title and image
+   * Adds a reference to a connection by origin. Ignores the 'metamask' origin.
+   * Caller must ensure that the returned id is stored such that the reference
+   * can be deleted later.
    *
-   * @param {*} outStream - The stream to provide the api over.
+   * @param {string} origin - The connection's origin string.
+   * @param {Object} options - Data associated with the connection
+   * @param {Object} options.engine - The connection's JSON Rpc Engine
+   * @returns {string} - The connection's id (so that it can be deleted later)
    */
-  setupPublicApi (outStream) {
-    const dnode = Dnode()
-    // connect dnode api to remote connection
-    pump(
-      outStream,
-      dnode,
-      outStream,
-      (err) => {
-        // report any error
-        if (err) {
-          log.error(err)
-        }
-      }
-    )
+  addConnection (origin, { engine }) {
+
+    if (origin === 'metamask') {
+      return null
+    }
 
-    const getRemote = createDnodeRemoteGetter(dnode)
+    if (!this.connections[origin]) {
+      this.connections[origin] = {}
+    }
 
-    const publicApi = {
-      // wrap with an await remote
-      getSiteMetadata: async () => {
-        const remote = await getRemote()
-        return await pify(remote.getSiteMetadata)()
-      },
+    const id = nanoid()
+    this.connections[origin][id] = {
+      engine,
+    }
+
+    return id
+  }
+
+  /**
+   * Deletes a reference to a connection, by origin and id.
+   * Ignores unknown origins.
+   *
+   * @param {string} origin - The connection's origin string.
+   * @param {string} id - The connection's id, as returned from addConnection.
+   */
+  removeConnection (origin, id) {
+
+    const connections = this.connections[origin]
+    if (!connections) {
+      return
+    }
+
+    delete connections[id]
+
+    if (Object.keys(connections.length === 0)) {
+      delete this.connections[origin]
+    }
+  }
+
+  /**
+   * Causes the RPC engines associated with the connections to the given origin
+   * to emit a notification event with the given payload.
+   * Does nothing if the extension is locked or the origin is unknown.
+   *
+   * @param {string} origin - The connection's origin string.
+   * @param {any} payload - The event payload.
+   */
+  notifyConnections (origin, payload) {
+
+    const { isUnlocked } = this.getState()
+    const connections = this.connections[origin]
+    if (!isUnlocked || !connections) {
+      return
     }
 
-    return publicApi
+    Object.values(connections).forEach(conn => {
+      conn.engine && conn.engine.emit('notification', payload)
+    })
   }
 
+  /**
+   * Causes the RPC engines associated with all connections to emit a
+   * notification event with the given payload.
+   * Does nothing if the extension is locked.
```

Comment:
The provider should be notified when the extension is locked, as it needs to emit an `accountsChanged` event in that case.

Maybe it makes sense for `notifyAllConnections` to do nothing if the extension is locked, but _something_ needs to notify the provider :thinking:

Reply:
The functionality is a little all over the place, but the provider gets state updates from the background and emits events when the extension becomes locked: https://github.com/MetaMask/metamask-inpage-provider/blob/master/index.js#L75-L93

Reply:
This is how we've been doing it, and changing it is probably breaking, but I'm wondering why we're exposing `networkVersion` and `chainId` even if the extension is locked: https://github.com/MetaMask/metamask-extension/blob/218a4221288909b65c2f2300ef14e337748a7e78/app/scripts/metamask-controller.js#L389-L395

Reply:
Ah I see now, thanks! This does indeed work.

I had thought I tested it and saw it not working, but I was misled by an minor unrelated bug in the `contract-test` dapp.

Reply:
The bug is that once the 'Connect' button is disabled, it doesn't get re-enabled again. I fixed it already in [this PR](https://github.com/MetaMask/test-dapp/pull/1) but forgot to update the extension too; I'll do that now.

Reply:
I suppose because you can call `eth_chainId` and `net_version` even if the extension is locked. And that being a different topic, we needn't worry 🤷‍♂ 

The interesting thing here is that, with `notifyConnections` and `notifyAllConnections`, I sense a way of getting rid of the detestable public config store entirely. We don't need to do it now, but I think we could!

Reply:
The `chainId` and `network`  issue is interesting 🤔 

It does seem wrong that we're allowing those updates to go through even while the extension is unlocked, but yes that would definitely be a breaking change. Maybe it's harmless, mostly? 🤷‍♂  Or maybe we can stop doing that some other time.

Reply:
The contract dapp fix is here: https://github.com/MetaMask/metamask-extension/pull/7587